fix: query GL_MAX_SAMPLES before setting multisampling hint

pandinocoder opened the pull request:
### Description of Change

- adds better located exceptions to catch things like "couldn't find matching glx visual"
- creates and disposes of temporary gl-enabled window and gl context to grab GL_MAX_SAMPLES from

This specifically fixes cases where `MultiSampleCount` is set to something like 16 on a build for Desktop GL and then is run on weaker hardware like a Steam Deck where `GL_MAX_SAMPLES` is 8.

Games do not have an API to query for this information to set a valid value for `MultiSampleCount` so the error is effectively unrecoverable and extremely hard to diagnose.

The exceptions currently thrown are only symptoms of the issue, they don't expose the actual failure which in this case needed to be queried for via `SDL_GetError()`.

### Contributor Declaration

- [x] I certify that no LLM's were used to generate any code or documentation in this contribution.

pandinocoder committed:
fix: query GL_MAX_SAMPLES before setting multisampling hint
- adds better located exceptions to catch things like "couldn't find matching glx visual"
- creates and disposes of temporary gl-enabled window and gl context to grab GL_MAX_SAMPLES from

mrhelmut commented:
You should be able to query ```GraphicsCapabilities.MaxMultiSampleCount``` in ```Game.Initialize()``` and apply graphics device parameters from there. Does that work?

DesktopGL comes with some limitation compared to WindowsDX in the fact that some graphics capabilities can be queried only after the Game constructor due to how SDL operates things under the hood. I'm a bit on the fence to create a fake window just to query that, I feel like it might introduce more issues in the long run that fixing this use case.

lodicolo commented:
`GraphicsCapabilities` and `GraphicsCapabilities.MaxMultiSampleCount` are internal so it would require some code changes to expose this.

It also has to create a temporary window no matter what, because GL can't query the API without an active context, and the way MonoGame is consuming SDL and GL it requires creating an SDL window.

I think creating the full graphics device chain in managed memory is creating a lot more garbage and introducing more opportunities for bugs instead of just quickly creating and destroying an SDL window in a controlled manner and then creating the real window. This is also, to my understanding, standard practice when using SDL (and GL).

I also personally have a concern with the approach of exposing this to the end users and expecting them to deal with it because:
- `MaxSamples` value is *not* always reliable, it reports 32 on my desktop when in reality it only supports up to 16 -- my Steam Deck reports correctly. Ordinarily I would be against my own fix for this, but the PR is meant as a "best effort" to prevent a crash, and includes better messaging so the end-user doesn't have to dive into dependencies with a debugger to figure out the root cause. I think having a potentially incorrect value as the maximum acceptable for a best effort, but not acceptable for a public API.
- Seems weird to expect users to know to poll graphics capabilities and then set the value
- Correct me if I am wrong, but in DX this can be done in the constructor, so this is also a GL-specific issue. I think it goes against the spirit of MonoGame to expect users to understand the internals of MonoGame and its dependencies in order to properly fix this issue.

dellis1972 commented:
After reading the docs on this, I think we can handle this a different way (I could be wrong though).
We can leave the code to call `SetAttribute` as it is now since it is only a `hint`. If the OpenGL driver cannot handle it is will use the maximum value it can. 
We should then use `GetAttribute` to query the actual value from the driver after the window has been created. Then update the presentationParameters.MultiSampleCount  accordingly.

Something like

```
if (presentationParameters.MultiSampleCount > 0)
{
    Sdl.GL.SetAttribute(Sdl.GL.Attribute.MultiSampleBuffers, 1);
    Sdl.GL.SetAttribute(Sdl.GL.Attribute.MultiSampleSamples, presentationParameters.MultiSampleCount);
}
((SdlGameWindow)SdlGameWindow.Instance).CreateWindow();
Sdl.GL.GetAttribute (Sdl.GL.Attribute.MultiSampleSamples, out int samples);
presentationParameters.MultiSampleCount = samples;
```

See https://wiki.libsdl.org/SDL2/SDL_GL_GetAttribute

mrhelmut commented:
> it will use the maximum value it can

We need to verify that it actually does that. But if it does without error, it seems relevant.

lodicolo commented:
> If the OpenGL driver cannot handle it is will use the maximum value it can.

Unfortunately the GL driver doesn't intelligently do this, and that causes the crash that this PR is meant to prevent.

> We can leave the code to call `SetAttribute` as it is now since it is only a `hint`.

It may be called a hint, but it's not treated as something optional. The `SetAttribute()` call with an invalid `GL_SAMPLES` value will cause the context creation that happens after it to fail. `GL_SAMPLES` (abstracted as `SDL_GL_MULTISAMPLESAMPLES`) must be set to a value supported by the driver before context creation, and changing it requires re-creating the context.

`Sdl.GL.GetAttribute()` would call `glGetIntegerv()` which cannot be used without a current valid GL context, which will fail to be created if `GL_MAX_SAMPLES` is set too high. The hidden dummy window being created would have the default `GL_MAX_SAMPLES` value of `0` which would allow it to be safely created in order to poll the `GL_MAX_SAMPLES` value in order to limit it.

Note that SDL2 does not expose `GL_MAX_SAMPLES` via `SDL_GL_GetAttribute()`, GL must be queried directly.

ThomasFOG commented:
There is already an hidden window created here that gets destroyed when the actual window is created (I don't know why it's created fullscreen, that seems weird but likely there for a reason):

https://github.com/MonoGame/MonoGame/blob/d5bf79ca793d1e36c4060e3a178a3f7e32c9653e/MonoGame.Framework/Platform/SDL/SDLGameWindow.cs#L133

This could be used to initialize the temporary OpenGL context by passing ``` Sdl.Window.State.OpenGL``` to it.
Remains the question of ```glGetIntegerv``` which we should avoid loading twice (but the problem is that entry point loading need an active GL context to function properly).

@pandinocoder could you look into this? Otherwise the PR will get stale and closed.

SimonDarksideJ commented:
Any update @pandinocoder / @ThomasFOG ?  Else this will be closed in 1 week as incomplete.
Also has merge conflicts now

pandinocoder committed:
Merge branch 'develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Add OpenGL to the window created during PlatformCreate(), and use that handle if possible to grab GL_MAX_SAMPLES

pandinocoder committed:
Modify MSAAEnabled so that it can work for DesktopGL

pandinocoder committed:
Merge remote-tracking branch 'upstream/develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
MonoGame.OpenGL should only be used if on DesktopGL

pandinocoder committed:
Merge remote-tracking branch 'upstream/develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Add [RunOnUI] to GameTest.FixtureBase.SetUp and TearDown to address failing tests

pandinocoder committed:
Merge remote-tracking branch 'upstream/develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Merge branch 'develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Revert [RunOnUI] since it seems to have worsened the test situation and window re-use

pandinocoder committed:
Add both `[RunOnUI]` and `[NonParallelizable]`

pandinocoder committed:
Merge remote-tracking branch 'upstream/develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
remove [NonParallelizable] (failing on Mac) and add [RunOnUI] to fixture (not tried yet)

pandinocoder committed:
Add [RunOnUI] to all failing fixtures

It is marked Inherited=true, but clearly there is some difference between these fixtures and GraphicsDeviceTestFixtureBase which is the only one that is not showing failures in the builds

pandinocoder committed:
Adding [NonParallizable] to the test classes but only for DESKTOPGL (made [RunOnUI] also be DESKTOPGL-only as well)

I noticed that GraphicsDeviceManagerTest is actually never run on any of the test runners, so I'm comparing the attributes on the failing tests to something that is supposed to be run like IndexBufferTest instead

pandinocoder committed:
Remove _handle check-and-throw-NoSuitableGraphicsDeviceException when the SDL window that is being created doesn't have OpenGL being loaded

pandinocoder committed:
Revert changes that are no longer necessary for the PR

pandinocoder committed:
Merge branch 'develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Merge branch 'develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Merge branch 'develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Merge branch 'develop' into fix-multisampling-no-matching-glx-visual

pandinocoder committed:
Merge branch 'develop' into fix-multisampling-no-matching-glx-visual
